src/extension.ts

```diff
@@ -25,9 +48,38 @@ export async function activate(context: vscode.ExtensionContext) {
   context.subscriptions.push(git.onDidOpenRepository(() => checkEnabled(git)));
   context.subscriptions.push(git.onDidCloseRepository(() => checkEnabled(git)));
 
+  // Create a debounced version of updateStatusBarItem
+  let updateTimeout: NodeJS.Timeout | null = null;
+  const debouncedUpdateStatusBar = (editor: vscode.TextEditor | undefined) => {
+    if (updateTimeout) {
+      clearTimeout(updateTimeout);
+    }
+    updateTimeout = setTimeout(() => {
+      updateStatusBarItem(editor);
+    }, 50); // 50ms debounce
+  };
+
+  // Watch for active editor changes to update icon state
+  context.subscriptions.push(
+    vscode.window.onDidChangeActiveTextEditor((editor) => {
+      debouncedUpdateStatusBar(editor);
+    })
+  );
+
+  // Set initial icon state
+  updateStatusBarItem(vscode.window.activeTextEditor);
+
+  // Initial check of enabled state
+  await checkEnabled(git);
+
+  // Watch for store changes
   reaction(
-    () => [store.enabled],
-    () => checkEnabled(git)
+    () => store.enabled,
+    (enabled) => {
+      updateContext(enabled, true);
+      checkEnabled(git);
+      debouncedUpdateStatusBar(vscode.window.activeTextEditor);
```

Comment:
I think this should be handled by the checkEnabled function, and effectively revert to the previous code. That way this reaction is simply watching for enabled changes and then updating the UX based on that (via checkEnabled)

Reply:
Actually, you already have a reaction in the watcher module that will update the status bar when the store.enabled changes. So I think we can just remove this?